### PR TITLE
Extending queries by modifing TxPaginationMeta

### PR DESCRIPTION
**Pull request by Pawlak00**
### Description of the Change
This is prove of concept for extending queries with time interval (from,to). Protobufs were modified and some basic logic was added. Validation of passed strings is not provided yet. Modified queries are getAccountTransactions and GetAccountAssetTransactions
Protobufs were modified, then neccesary getters were added.
Passed parameters are attached to sql query in executeTransactionsQuery, by applyer lamba function,
created separately in each function

### Benefits

 What benefits will be realized by the code change? 
Allowing to query for Transactions in given time interval and beetwen two given block heights
### Possible Drawbacks 

What are the possible side-effects or negative impacts of the code change? 
Usage of many if's in GetAccountTransactions/GetAccountAssetTransactions in postgres_specific_query_executor.cpp

### Usage Examples or Tests *[optional]*
By rebuilding python library with protbuffs from code you can use it in Python library

### Alternate Designs *[optional]*

Usage of different types for passing data as timestamp was considered, but I decided to use string, as it is not needed to cast from int to string

### CODE REVIEW

**Comment by baziorek** [comment]
There is reply from Iroha Core Team:
image
It looks like we have to change string format to protobuf datetime format.

File: shared_model/backend/protobuf/queries/impl/proto_tx_pagination_meta.cpp
Comment (baziorek) [inline shared_model/backend/protobuf/queries/impl/proto_tx_pagination_meta.cpp:12]: exactly:)

File: shared_model/backend/protobuf/queries/impl/proto_tx_pagination_meta.cpp
Comment (baziorek) [inline shared_model/backend/protobuf/queries/impl/proto_tx_pagination_meta.cpp:44]: `TimestampToMicroseconds` instead of `TimestampToNanoseconds`

File: shared_model/backend/protobuf/queries/impl/proto_tx_pagination_meta.cpp
Comment (baziorek) [inline shared_model/backend/protobuf/queries/impl/proto_tx_pagination_meta.cpp:48]: spaces around `::` to remove

File: shared_model/backend/protobuf/queries/proto_get_account_asset_transactions.hpp
Comment (baziorek) [inline shared_model/backend/protobuf/queries/proto_get_account_asset_transactions.hpp:38]: I suggest to remove changes in the file (no changes other than removing newline)

File: shared_model/validators/validators_common.cpp
Comment (baziorek) [inline shared_model/validators/validators_common.cpp:9]:
```suggestion
```

**Review by baziorek — CHANGES_REQUESTED** [review]

File: irohad/ametsuchi/impl/postgres_specific_query_executor.cpp
Comment (Pawlak00) [reply irohad/ametsuchi/impl/postgres_specific_query_executor.cpp:744]: this is the thing I mean asking you priv

File: shared_model/backend/protobuf/queries/impl/proto_tx_pagination_meta.cpp
Comment (Pawlak00) [reply shared_model/backend/protobuf/queries/impl/proto_tx_pagination_meta.cpp:44]: I think it should actually be TimestampToMilliseconds. Look here : https://iroha.readthedocs.io/en/main/concepts_architecture/glossary.html#transaction

File: irohad/ametsuchi/impl/postgres_specific_query_executor.cpp
Comment (Pawlak00) [reply irohad/ametsuchi/impl/postgres_specific_query_executor.cpp]: solved by using `soci::use(first_tx_time, ind, "first_tx_time")`

**Review by baziorek — CHANGES_REQUESTED** [review]
Global changes to correct:
- Please don't change whitecharacters of lines which You don't change
- Please don't remove/add newlines at the end of existing files
- Please don't add/remove newlines between lines which You don't change

File: shared_model/backend/protobuf/queries/impl/proto_tx_pagination_meta.cpp
Comment (baziorek) [reply shared_model/backend/protobuf/queries/impl/proto_tx_pagination_meta.cpp]: You are right, sorry

File: irohad/ametsuchi/impl/postgres_specific_query_executor.cpp
Comment (Pawlak00) [reply irohad/ametsuchi/impl/postgres_specific_query_executor.cpp:739]: `soci::use(first_tx_time, ind, "first_tx_time")` is passing i_null to query if first_tx_time is std::nullopt. It is then changes by `:first_tx_time::text` to `NULL`. The if statement differs from else in first_hash in return statement

File: .gitignore
Comment (Pawlak00) [reply .gitignore:38]: file is same as in iroha main

File: docs/source/requirements.txt
Comment (Pawlak00) [reply docs/source/requirements.txt:11]: I copied this file from main, but is valid version on Jinja,

File: irohad/ametsuchi/impl/postgres_specific_query_executor.cpp
Comment (baziorek) [reply irohad/ametsuchi/impl/postgres_specific_query_executor.cpp:738]: OK

File: irohad/ametsuchi/impl/postgres_specific_query_executor.cpp
Comment (baziorek) [inline irohad/ametsuchi/impl/postgres_specific_query_executor.cpp:1515]: trailing space added, it should be removed

File: shared_model/backend/protobuf/queries/impl/proto_get_account_transactions.cpp
Comment (baziorek) [inline shared_model/backend/protobuf/queries/impl/proto_get_account_transactions.cpp:30]: trailing space

File: shared_model/backend/protobuf/queries/proto_get_account_transactions.hpp
Comment (baziorek) [inline shared_model/backend/protobuf/queries/proto_get_account_transactions.hpp:10]: I suggest to restore the file to previous version - restore order of includes

File: shared_model/builders/protobuf/builder_templates/query_template.hpp
Comment (baziorek) [inline shared_model/builders/protobuf/builder_templates/query_template.hpp:225]: the comment looks like shout be removed

File: shared_model/interfaces/queries/get_account_transactions.hpp
Comment (baziorek) [inline shared_model/interfaces/queries/get_account_transactions.hpp:28]: please, revert changes from the file, because only newline is removed

File: shared_model/validators/validators_common.hpp
Comment (baziorek) [inline shared_model/validators/validators_common.hpp:54]: I suggest to add information what `height` e.g. `block height`

**Review by baziorek — CHANGES_REQUESTED** [review]
There are still changes in whitespaces/newline removing in files with not changed code. Please correct its (remove Your changes from the files). Also it would be good not to change parts of files not around changed code.

File: shared_model/builders/protobuf/builder_templates/query_template.hpp
Comment (baziorek) [inline shared_model/builders/protobuf/builder_templates/query_template.hpp:138]: the code is duplicated twice, I suggest to extract method/function to setup timestamp from tx_time. I also suggest to change the magic numbers 1000 and 1000000 into `milliseconds_per_second`, similar in another case

File: test/module/irohad/ametsuchi/postgres_query_executor_test.cpp
Comment (baziorek) [inline test/module/irohad/ametsuchi/postgres_query_executor_test.cpp:709]: The loop is the same here and in lines [697, 702] - why not to extract function/method?

File: test/module/irohad/ametsuchi/postgres_query_executor_test.cpp
Comment (baziorek) [inline test/module/irohad/ametsuchi/postgres_query_executor_test.cpp:1632]: Some tests are missing - You are testing 10 transactions and filtering transactions before all of them and after all of them with first_tx_time and last_tx_time - it is working. But it is not testing if Your code is really working. So I suggest that test:
1. Commit 3 transactions
2. Set first_tx_time
3. Commit 4 transactions
4. set last_tx_time
5. Commit 5 transactions
6. As a result You should have 4 transactions.

**Review by baziorek — CHANGES_REQUESTED** [review]
Code is almost ready. Just few small corrections and adding the test I've mentioned in the review.
The next thing is to update documentation. Then please let me know.
If the next version would be good we can ask Iroha-Core-Team to take a look into the code.

File: test/module/irohad/ametsuchi/postgres_query_executor_test.cpp
Comment (Pawlak00) [reply test/module/irohad/ametsuchi/postgres_query_executor_test.cpp:1686]: I  solved it by passing first and last time by reference, and than using it in query. I also have to sleep the thread for few miliseconds to get different timestamps of transactions

**Comment by kuvaldini** [comment]
Please 
`git rebase support/1.2.x --signof`

File: test/module/irohad/ametsuchi/postgres_query_executor_test.cpp
Comment (baziorek) [inline test/module/irohad/ametsuchi/postgres_query_executor_test.cpp:678]: Despite the fact that the code is in tests I suggest to use const reference if possible

File: test/module/irohad/ametsuchi/postgres_query_executor_test.cpp
Comment (baziorek) [inline test/module/irohad/ametsuchi/postgres_query_executor_test.cpp:683]: Despite the fact that the code is in tests I suggest to use const reference if possible

File: test/module/irohad/ametsuchi/postgres_query_executor_test.cpp
Comment (baziorek) [reply test/module/irohad/ametsuchi/postgres_query_executor_test.cpp:1686]: OK

**Review by baziorek — APPROVED** [review]
Small changes, but rather IMO it looks good.
Now correct suggestions from iroha core team. Then change from Draft into normal PR. Then we can ask Iroha Core Team on chat for review.

File: docs/source/develop/api/queries.rst
Comment (kuvaldini) [inline docs/source/develop/api/queries.rst:77]: These breaks current API. We should save new API compatible with older clients. New parameters should be optional and ordering (numbers of filds shouldd stay the same)

**Review by kuvaldini — CHANGES_REQUESTED** [review]

File: docs/source/develop/api/queries.rst
Comment (Pawlak00) [reply docs/source/develop/api/queries.rst:77]: You mean to place them after ordering parameter?

File: docs/source/develop/api/queries.rst
Comment (kuvaldini) [inline docs/source/develop/api/queries.rst:61]: Please DRY. Let this code-block be sourced from original queries.proto. Let future changes to queries.proto be synchronized automatically.
Looks like you'll need to move `TxPaginationMeta`, please consider ways to do.

File: docs/source/develop/api/queries.rst
Comment (kuvaldini) [reply docs/source/develop/api/queries.rst:77]: I mean number after assign must be saved the same 
```suggestion
      Ordering ordering = 3;
```

File: docs/source/develop/api/queries.rst
Comment (Pawlak00) [reply docs/source/develop/api/queries.rst]: I think that other code fields in this .rst file could be done this way. Should I modify them also?

File: test/module/shared_model/validators/protobuf/proto_query_validator_test.cpp
Comment (baziorek) [inline test/module/shared_model/validators/protobuf/proto_query_validator_test.cpp:146]: I suggest to remove the commented code

**Review by baziorek — APPROVED** [review]
Code looks really good now. So I'm satisfied with the code, now it is time to satisfy Iroha Team:).

File: docs/source/develop/api/queries.rst
Comment (baziorek) [reply docs/source/develop/api/queries.rst]: Sorry for breaking in: Changing rest of fields in the .rst file would be not hard, but it would make PR harder to verify and bigger. So I suggest not to change other fields in .rst file in the PR.

The pointed issue is corrected for TxPaginationMeta so I'm marking as resolved.

File: docs/source/develop/api/queries.rst
Comment (baziorek) [reply docs/source/develop/api/queries.rst]: It is restored, so I'm marking its as resolved

**Comment by kuvaldini** [comment]
Did you check this PR does not break existing API?

**Comment by kamilsa** [comment]
@Pawlak00 can you please rebase your commits on the latest support/1.2.x?

**Comment by kamilsa** [comment]
> Did you check this PR does not break existing API?

@kuvaldini I think it does not. Fields added queries.proto are optional and order is preserved

**Comment by baziorek** [comment]
Something happend during rebasing from:
`main` branch into `support/1.2.x`

**Comment by kamilsa** [comment]
> Something happend during rebasing from:
> `main` branch into `support/1.2.x`

@Pawlak00 I think the easiest thing to remove redundant commits is just squash everything into a single commit. Otherwise there are no major issues

File: shared_model/validators/protobuf/proto_query_validator.cpp
Comment (iceseer) [inline shared_model/validators/protobuf/proto_query_validator.cpp:47]: + first_tx_time <= last_tx_time, if both are present

File: shared_model/validators/protobuf/proto_query_validator.cpp
Comment (iceseer) [inline shared_model/validators/protobuf/proto_query_validator.cpp:63]: + first_tx_height <= last_tx_height, if both are present

**Review by iceseer — CHANGES_REQUESTED** [review]

**Comment by kuvaldini** [comment]
Please make sure you include only your own changes to this PR.

It would be the best approach to squash all in one commit.

I suggest
```
cd iroha
git checkout origin/support/1.2.x -b new-tx-pagination-meta
git checkout Pawlak00/main -- .
git add .....................
git commit -m"[client-api] GOOD short with exhaustive description commit message"
```
This need to be **carefully validated**.

These commits must be removed from PR.
изображение

**Comment by kuvaldini** [comment]
@Pawlak00 ^^

If need help, You can reach me for help in telegram for instant messaging https://t.me/kuvaldini

cc @baziorek

**Comment by Pawlak00** [comment]
I messed up with one test, so it is not ready to merge now, hope i will figure it out today

**Comment by kuvaldini** [comment]
Good luck. 
It is very good test failed before merge, not after

**Comment by Pawlak00** [comment]
Issue is solved, so it is ready to proceed @kuvaldini @baziorek

**Review by kuvaldini — APPROVED** [review]

**Review by iceseer — APPROVED** [review]

**Review by kuvaldini — APPROVED** [review]

**Pull request merged**
